src/with_clause_parser.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+#include <postgres.h>
+#include <c.h>
+
+#include <nodes/parsenodes.h>
+
+#define TS_WITH_DESERIALIZE_ERROR(def, value, type) \
+	ereport(ERROR, \
+		(errcode(ERRCODE_INVALID_PARAMETER_VALUE), \
+			errmsg("invalid value for %s.%s '%s'", \
+				def->defnamespace, def->defname, value), \
+			errhint("%s.%s must be a %s", def->defnamespace, def->defname, type)))
+
+#define TS_WITH_DUPLICATE_PARAMETER_ERROR(def) \
+    ereport(ERROR, \
+            (errcode(ERRCODE_AMBIGUOUS_PARAMETER), \
+                errmsg("duplicate parameter \"%s.%s\"", \
+                    def->defnamespace, def->defname))); \
+
+/* find the length of a statically sized array */
+#define TS_ARRAY_LEN(array) (sizeof(array) / sizeof(*array))
+
+typedef struct WithClauseArg
+{
+	const char *arg_name;
+	Datum		(*deserializer) (const DefElem *element, const char *value, void *state);
```

Comment:
maybe we can provide a pg function for all the deserializations `boolin`, `int4in`, `namein` etc?